{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from distributed import Client, LocalCluster\n",
    "from dask_jobqueue import SGECluster\n",
    "import numpy as np\n",
    "from plasp.config import SAVE_DIR\n",
    "from plasp.dataloader import MULANLoader, LIBSVMLoader, Synthesizer, FoldsLoader\n",
    "from plasp.scripts.grid_search import iterate_on_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def experiment(dataset, Loader, methods):\n",
    "    datasets = [dataset]\n",
    "    corruptions = [i / 10 for i in range(10)]\n",
    "    nb_folds = 10\n",
    "    floader = FoldsLoader(Loader, cross_on_weak_loss=False)\n",
    "    C_sigmas = np.logspace(-2, 2, 5)\n",
    "    C_lambdas = np.logspace(-3, 3, 7)\n",
    "    np.random.seed(0)\n",
    "    res_dict = iterate_on_datasets(datasets, corruptions, nb_folds, floader,\n",
    "                                   methods, C_sigmas, C_lambdas, low_rank=200)\n",
    "    if floader.wk:\n",
    "        dataset += '_weak'\n",
    "    with open(SAVE_DIR / (dataset + '.pk'), 'wb') as f:\n",
    "        pickle.dump(res_dict, f)\n",
    "    return res_dict\n",
    "\n",
    "\n",
    "def MULAN_experiment(dataset):\n",
    "    methods = ['plasp', 'low rank', 'diffusion']\n",
    "    return experiment(dataset, MULANLoader, methods)\n",
    "\n",
    "\n",
    "def LIBSVM_experiment(dataset):\n",
    "    methods = ['plasp', 'averaging', 'diffusion']\n",
    "    return experiment(dataset, LIBSVMLoader, methods)\n",
    "\n",
    "\n",
    "def synthetic_experiment(dataset):\n",
    "    methods = ['plasp', 'diffrac', 'diffusion', 'low rank']\n",
    "    return experiment(dataset, Synthesizer, methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dask_cluster(on_server, n_workers):\n",
    "    if on_server:\n",
    "        env_extra = [\n",
    "            'conda activate main',\n",
    "            'source /home/vcabanne/.bashrc',\n",
    "            'export LANG=en_US.UTF-8',\n",
    "            'export LC_ALL=en_US.UTF-8'\n",
    "        ]\n",
    "        env_extra = [\n",
    "            '$ -e /sequoia/data2/vcabanne/logs',\n",
    "            '$ -o /sequoia/data2/vcabanne/logs',\n",
    "        ] + env_extra\n",
    "\n",
    "        local_directory = '/sequoia/data2/vcabanne/dask-spill'\n",
    "\n",
    "        cluster = SGECluster(queue='all.q,bigmem.q',\n",
    "                             resource_spec='h_vmem=12000M,mem_req=8000M',\n",
    "                             walltime='720:00:00',\n",
    "                             name='plasp',\n",
    "                             cores=1,\n",
    "                             memory='8000m',\n",
    "                             processes=1,\n",
    "                             interface='en0',\n",
    "                             env_extra=env_extra,\n",
    "                             local_directory=local_directory)\n",
    "        cluster.scale(n_workers)\n",
    "\n",
    "    else:\n",
    "        cluster = LocalCluster(n_workers=n_workers, threads_per_worker=1,\n",
    "                               processes=False)\n",
    "\n",
    "    return cluster\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished coro=<_wrap_awaitable() done, defined at /Users/viviencabannes/miniconda3/envs/main/lib/python3.7/asyncio/tasks.py:623> exception=FileNotFoundError(2, \"No such file or directory: 'qsub'\")>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/asyncio/tasks.py\", line 630, in _wrap_awaitable\n",
      "    return (yield from awaitable.__await__())\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py\", line 50, in _\n",
      "    await self.start()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 310, in start\n",
      "    out = await self._submit_job(fn)\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 293, in _submit_job\n",
      "    return self._call(shlex.split(self.submit_command) + [script_filename])\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 381, in _call\n",
      "    cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/subprocess.py\", line 800, in __init__\n",
      "    restore_signals, start_new_session)\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/subprocess.py\", line 1551, in _execute_child\n",
      "    raise child_exception_type(errno_num, err_msg, err_filename)\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'qsub': 'qsub'\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <zmq.eventloop.ioloop.ZMQIOLoop object at 0x1a24a6d7d0>>, <Task finished coro=<SpecCluster._correct_state_internal() done, defined at /Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py:300> exception=FileNotFoundError(2, \"No such file or directory: 'qsub'\")>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/tornado/ioloop.py\", line 743, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/tornado/ioloop.py\", line 767, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py\", line 335, in _correct_state_internal\n",
      "    await w  # for tornado gen.coroutine support\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py\", line 50, in _\n",
      "    await self.start()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 310, in start\n",
      "    out = await self._submit_job(fn)\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 293, in _submit_job\n",
      "    return self._call(shlex.split(self.submit_command) + [script_filename])\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 381, in _call\n",
      "    cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/subprocess.py\", line 800, in __init__\n",
      "    restore_signals, start_new_session)\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/subprocess.py\", line 1551, in _execute_child\n",
      "    raise child_exception_type(errno_num, err_msg, err_filename)\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'qsub': 'qsub'\n",
      "Task exception was never retrieved\n",
      "future: <Task finished coro=<_wrap_awaitable() done, defined at /Users/viviencabannes/miniconda3/envs/main/lib/python3.7/asyncio/tasks.py:623> exception=FileNotFoundError(2, \"No such file or directory: 'qsub'\")>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/asyncio/tasks.py\", line 630, in _wrap_awaitable\n",
      "    return (yield from awaitable.__await__())\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py\", line 50, in _\n",
      "    await self.start()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 310, in start\n",
      "    out = await self._submit_job(fn)\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 293, in _submit_job\n",
      "    return self._call(shlex.split(self.submit_command) + [script_filename])\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/dask_jobqueue/core.py\", line 381, in _call\n",
      "    cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/subprocess.py\", line 800, in __init__\n",
      "    restore_signals, start_new_session)\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/subprocess.py\", line 1551, in _execute_child\n",
      "    raise child_exception_type(errno_num, err_msg, err_filename)\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'qsub': 'qsub'\n"
     ]
    },
    {
     "ename": "CancelledError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCancelledError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-6ab978eadf3d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     40\u001b[0m             'ML-1000', 'ML-400', 'ML-100']\n\u001b[1;32m     41\u001b[0m \u001b[0mfutures\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msynthetic_experiment\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatasets\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 42\u001b[0;31m \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfutures\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/main/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36mgather\u001b[0;34m(self, futures, errors, direct, asynchronous)\u001b[0m\n\u001b[1;32m   1892\u001b[0m                 \u001b[0mdirect\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdirect\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1893\u001b[0m                 \u001b[0mlocal_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlocal_worker\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1894\u001b[0;31m                 \u001b[0masynchronous\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0masynchronous\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1895\u001b[0m             )\n\u001b[1;32m   1896\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/main/lib/python3.7/site-packages/distributed/client.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(self, func, asynchronous, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    776\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    777\u001b[0m             return sync(\n\u001b[0;32m--> 778\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    779\u001b[0m             )\n\u001b[1;32m    780\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/main/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36msync\u001b[0;34m(loop, func, callback_timeout, *args, **kwargs)\u001b[0m\n\u001b[1;32m    346\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    347\u001b[0m         \u001b[0mtyp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 348\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    349\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    350\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/main/lib/python3.7/site-packages/distributed/utils.py\u001b[0m in \u001b[0;36mf\u001b[0;34m()\u001b[0m\n\u001b[1;32m    330\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcallback_timeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    331\u001b[0m                 \u001b[0mfuture\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0masyncio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait_for\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfuture\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallback_timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 332\u001b[0;31m             \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32myield\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    333\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    334\u001b[0m             \u001b[0merror\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexc_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/main/lib/python3.7/site-packages/tornado/gen.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    733\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    734\u001b[0m                     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 735\u001b[0;31m                         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    736\u001b[0m                     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    737\u001b[0m                         \u001b[0mexc_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexc_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCancelledError\u001b[0m: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <zmq.eventloop.ioloop.ZMQIOLoop object at 0x1a24863bd0>>, <Task finished coro=<SpecCluster._close() done, defined at /Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py:373> exception=AttributeError(\"'NoneType' object has no attribute 'close'\")>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/tornado/ioloop.py\", line 743, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/tornado/ioloop.py\", line 767, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py\", line 386, in _close\n",
      "    await self.scheduler_comm.close(close_workers=True)\n",
      "AttributeError: 'NoneType' object has no attribute 'close'\n",
      "tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <zmq.eventloop.ioloop.ZMQIOLoop object at 0x1a217de3d0>>, <Task finished coro=<SpecCluster._close() done, defined at /Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py:373> exception=AttributeError(\"'NoneType' object has no attribute 'close'\")>)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/tornado/ioloop.py\", line 743, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/tornado/ioloop.py\", line 767, in _discard_future_result\n",
      "    future.result()\n",
      "  File \"/Users/viviencabannes/miniconda3/envs/main/lib/python3.7/site-packages/distributed/deploy/spec.py\", line 386, in _close\n",
      "    await self.scheduler_comm.close(close_workers=True)\n",
      "AttributeError: 'NoneType' object has no attribute 'close'\n",
      "distributed.client - ERROR - Failed to reconnect to scheduler after 10.00 seconds, closing client\n",
      "_GatheringFuture exception was never retrieved\n",
      "future: <_GatheringFuture finished exception=CancelledError()>\n",
      "concurrent.futures._base.CancelledError\n"
     ]
    }
   ],
   "source": [
    "# architecture\n",
    "on_server = True\n",
    "n_workers = 1\n",
    "cluster = get_dask_cluster(on_server, n_workers)\n",
    "client = Client(cluster)\n",
    "\n",
    "# MULAN data\n",
    "MULAN_datasets = MULANLoader.datasets\n",
    "MULAN_datasets = [\n",
    "    'birds',\n",
    "    'CAL500',\n",
    "    'emotions',\n",
    "    'enron',\n",
    "    'flags',\n",
    "    'genbase',\n",
    "    'medical',\n",
    "    'scene',\n",
    "    'yeast'\n",
    "]\n",
    "futures = client.map(MULAN_experiment, MULAN_datasets)\n",
    "# results = client.gather(futures)\n",
    "\n",
    "# LIBSSM data\n",
    "LIBSVM_datasets = LIBSVMLoader.datasets\n",
    "LIBSVM_datasets = [\n",
    "    'iris',\n",
    "    'wine',\n",
    "    'svmguide2',\n",
    "    'glass',\n",
    "    'svmguide4',\n",
    "    'vehicle',\n",
    "    'vowel',\n",
    "    'dna'\n",
    "]\n",
    "futures = client.map(LIBSVM_experiment, LIBSVM_datasets)\n",
    "# results = client.gather(futures)\n",
    "\n",
    "# Synthetic data\n",
    "datasets = ['CL-1000', 'CL-400', 'CL-100',\n",
    "            'ML-1000', 'ML-400', 'ML-100']\n",
    "futures = client.map(synthetic_experiment, datasets)\n",
    "results = client.gather(futures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
